Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1801 @@
+                        var objIVBProjectExPtr = Marshal.GetComInterfaceForObject(_wrappedObject, typeof(IVBEProject));
+                        _cachedIVBProjectEx2 = new RestrictComInterfaceByAggregation<IVBEProject2>(objIVBProjectExPtr - IntPtr.Size, queryForType: false);
+                    }
+                    else
+                    {
+                        throw new ArgumentException("This ITypeLib is not hosted by the VBE, so does not support _IVBProjectEx");
+                    }
+                }
+
+                return (IVBEProject2)_cachedIVBProjectEx2.WrappedObject;
+            }
+        }*/
+
+        public static TypeLibWrapper FromVBProject(IVBProject vbProject)
+        {
+            using (var references = vbProject.References)
+            {
+                // Now we've got the references object, we can read the internal object structure to grab the ITypeLib
+                var internalReferencesObj = StructHelper.ReadComObjectStructure<VBEReferencesObj>(references.Target);
+                return new TypeLibWrapper(internalReferencesObj.TypeLib);
+            }
+        }
+
+        private void InitCommon()
+        {
+            TypeInfos = new TypeInfosCollection(this);
+            target_IVBEProject = target_ITypeLib as IVBEProject;
+            if (HasVBEExtensions) VBEReferences = new ReferencesCollection(this);
+        }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="rawObjectPtr">The raw unamanaged ITypeLib pointer</param>
+        public TypeLibWrapper(IntPtr rawObjectPtr)
+        {
+            if (!UnmanagedMemHelper.ValidateComObject(rawObjectPtr))
+            {
+                throw new ArgumentException("Expected COM object, but validation failed.");
+            };
+            target_ITypeLib = (ComTypes.ITypeLib)Marshal.GetObjectForIUnknown(rawObjectPtr);
+            Marshal.Release(rawObjectPtr);         // target_ITypeLib holds a reference to this now
+            InitCommon();
+        }
+
+        public TypeLibWrapper(ComTypes.ITypeLib rawTypeInfo)
+        {
+            target_ITypeLib = rawTypeInfo;
+            _wrappedObjectIsWeakReference = true;
+            InitCommon();
+        }
+
+        private bool _isDisposed;
+        public void Dispose()
+        {
+            if (_isDisposed) return;
+            _isDisposed = true;
+
+            _typeInfosWrapped?.Dispose();
+            if (!_wrappedObjectIsWeakReference) Marshal.ReleaseComObject(target_ITypeLib);
+        }
+
+        public TypeInfoWrapper GetSafeTypeInfoByIndex(int index)
+        {
+            IntPtr typeInfoPtr = IntPtr.Zero;
+            // We cast to our IVBETypeLib interface in order to work with the raw IntPtr for aggregation
+            ((ITypeLib_Ptrs)target_ITypeLib).GetTypeInfo(index, out typeInfoPtr);
+            var outVal = new TypeInfoWrapper(typeInf
```

Comment:
Same comment RE: wrapping in a `#if DEBUG`

Reply:
sure, thanks